FreeRTOS/FreeRTOS-Kernel
# Duplicate #625 : Fix xTaskNotifyWait & ulTaskNotifyTake determinism

Description
-----------
Fixes the bug described in this issue:
https://github.com/FreeRTOS/FreeRTOS-Kernel/issues/612

One unorthodox thing being done here, which I feel should be pointed out, is that I am calling vTaskSuspendAll() from inside a critical section. This is done to minimize the performance impact of this change, since one of the main benefits of task notifications is that they are faster than the other primitives. The only worry I see here is that it may cause issues with certain ports if they do tricky things inside portSOFTWARE_BARRIER() or portMEMORY_BARRIER(), but I cannot think of any examples.

Test Steps
-----------
The code seems to work on an RP2040.

Checklist:
----------



Related Issue
-----------
#625 


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
